### ISSUE TITLE
Add ability to call OCCA functions not just kernels from API

### ISSUE THREAD

**Issue by pdhahn** [libocca/occa#220]

**Comment by dmed256** [comment]
We could try to relax the restrictions for individual modes

- **Serial**: No need for `@outer` / `@inner`
- **OpenMP**: No need for `@inner`
- **CUDA/OpenCL/HIP**: Needs `@outer` / `@inner`

In that case maybe `@function` like you suggested for functions we want to "export" / make JIT-able. Would that satisfy your use-case?

**Comment by pdhahn** [comment]
I see what you're saying. I think what you are alluding to is probably too powerful for my needs.

For me, the objective is about keeping more of related logic together in the same `.okl` file if possible. Frankly, my use case is conveniently computing sizes of device memory allocations for kernel args (return value and work arrays). A number of my kernels take parameters that affect format and extent of returned result data. All I really would like is the ability to invoke a jit-able kind of function in the `.okl` file that runs on the host synchronously when invoked in application code using the OCCA API, no matter the OCCA mode. And propagate the return value of the function, too. Not anything more than that, really.

So in other words, OCCA has a very nice API that does jit compilation plus hidden dlopen/dlsym when invoking kernels running async on a device. Please extend the API to handle jit compilation plus dlopen/dlsym on functions running sync on the host, no matter what the mode, with code located in the `.okl` file.

**Comment by pdhahn** [comment]
My original title for this issue was incorrect as it referred to "OCCA function", so I changed it to pertain to "host function". In v0.2, "OCCA function" would mean a function that runs on the device, declared "occaFunction". Sorry about that.

**Comment by kris-rowe** [comment]
## References

- [CUDA: cuLaunchHostFunc](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__EXEC.html#group__CUDA__EXEC_1gab95a78143bae7f21eebb978f91e7f3f)
- [HIP: hipLaunchHostFunc](https://rocm.docs.amd.com/projects/HIP/en/latest/reference/hip_runtime_api/modules/launch_api.html#_CPPv417hipLaunchHostFunc11hipStream_t11hipHostFn_tPv)
- [SYCL: host tasks](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:interfaces.hosttasks)
- [OpenCL: clSetEventCallback](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clSetEventCallback)

**Comment by kris-rowe** [comment]
@sejalgupta11 Here is a SYCL example of calling a function pointer from a host task in that should roughly equate to what we need for that backend.

```
#include <iostream>
#include <sycl/sycl.hpp>
#include <vector>

namespace {

void axpy(int N, int alpha, const int* x, int* y) {
  for (int n = 0; n < N; ++n) 
    y[n] += alpha * x[n];
}

}  // namespace

int main(void) {
  sycl::queue sycl_queue{};

  const int N = 8192;
  auto* x = new int[N];
  auto* y = new int[N];

  for (int n = 0; n < N; ++n) {
    x[n] = 1; y[n] = 1;
  }

  const int alpha = 1;
  void (*axpy_pointer)(int, int, const int*, int*) = &axpy;

  sycl_queue.submit([&](sycl::handler& h) {
    // Check if capture-by-reference is needed below
    h.host_task( [=]() { axpy_pointer(N, alpha, x, y); });
  });
  sycl_queue.wait();

  for (int n = 0; n < N; ++n) {
    if (2 != y[n]) std::cout << "n: " << n << ", expected: 2, actual: " << y[n] << std::endl;
  }

  delete[] x;
  delete[] y;

  return 0;
}
```
